### PR TITLE
Require xcode_backend.sh arguments (introduced in Flutter v0.0.7)

### PR DESCRIPTION
**Pull request by jmagman**
## Description

xcode_backend.sh supported pre-Flutter v0.0.7 created apps that were missing a post-compilation build phase.  https://github.com/flutter/flutter/pull/64988 reland will require this build phase hook to change the bundled app's Info.plist.

Stop supporting the pre-v0.0.7 Xcode projects missing build phase.  Generate a build error on the slim chance these still exist in the wild:
Screen Shot 2020-09-02 at 1 31 16 PM

## Related Issues

#7913 
#64988

## Tests

I added the following tests:

Update the benchmark integration test Xcode projects that were created pre-v0.0.7.
https://github.com/flutter/flutter/pull/64988 postsubmit failed on the devicelab:
```
2020-09-02T11:44:03.947711: stdout:             ♦ codesign --force --verbose --sign A5CFF7959BD4B17F97127C518E8504F2D9562490 -- /Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Frameworks/Flutter.framework/Flutter
stdout:             /Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Frameworks/Flutter.framework/Flutter: signed bundle with Mach-O universal (armv7 x86_64 arm64) [io.flutter.flutter]
stdout:             /Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Info.plist: file does not exist or is not readable or is not a regular file (Error Domain=NSCocoaErrorDomain Code=260 "The file “Info.plist” couldn’t be opened because there is no such file." UserInfo={NSFilePath=/Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Info.plist, NSUnderlyingError=0x7fe71440c460 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}})
stdout:             ♦ plutil -insert NSBonjourServices -json ["_dartobservatory._tcp"] /Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Info.plist
stdout:             /Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Info.plist: file does not exist or is not readable or is not a regular file (Error Domain=NSCocoaErrorDomain Code=260 "The file “Info.plist” couldn’t be opened because there is no such file." UserInfo={NSFilePath=/Users/flutter/.cocoon/flutter/dev/benchmarks/complex_layout/build/ios/Profile-iphoneos/Runner.app/Info.plist, NSUnderlyingError=0x7fc1a7c0ce70 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}})
stdout:             Command PhaseScriptExecution failed with a nonzero exit code
```
These tests now pass.

## Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

File: packages/flutter_tools/bin/xcode_backend.sh [hunk 1]
```diff
@@ -315,9 +315,9 @@ EmbedAndThinFrameworks() {
 
 # Main entry point.
 if [[ $# == 0 ]]; then
-  # Backwards-compatibility: if no args are provided, build and embed.
-  BuildApp
-  EmbedFlutterFrameworks
+  # Named entry points were introduced in Flutter v0.0.7.
+  echo "error: Your Xcode project is incompatible with this version of Flutter. Run \"rm -rf ios/Runner.xcodeproj\" and \"flutter create .\" to regenerate."
+  exit -1
```
Comment (jmagman) [inline packages/flutter_tools/bin/xcode_backend.sh:320]: This is the actual change.  The rest of the PR updates the benchmark Xcode projects.

File: dev/benchmarks/complex_layout/ios/Runner.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -166,18 +180,33 @@
...
 			files = (
 			);
 			inputPaths = (
 			);
+			name = "Thin Binary";
 			outputPaths = (
 			);
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "/bin/sh $FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh";
```
Comment (jmagman) [inline dev/benchmarks/complex_layout/ios/Runner.xcodeproj/project.pbxproj:180]: This was what caused the devicelab errors on #64988 post-submit.  `xcode_backend.sh` without any args was building, thinning, and embedding.  However, at this point the generated Info.plist that we will need to edit for https://github.com/flutter/flutter/pull/64988 doesn't exist, so it was failing.

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/bin/xcode_backend.sh:320]: Wow!

**Review by dnfield — APPROVED** [review]
LGTM

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: packages/flutter_tools/bin/xcode_backend.sh [hunk 3]
```diff
@@ -315,9 +315,9 @@ EmbedAndThinFrameworks() {
 
 # Main entry point.
 if [[ $# == 0 ]]; then
-  # Backwards-compatibility: if no args are provided, build and embed.
-  BuildApp
-  EmbedFlutterFrameworks
+  # Named entry points were introduced in Flutter v0.0.7.
+  EchoError "error: Your Xcode project is incompatible with this version of Flutter. Run \"rm -rf ios/Runner.xcodeproj\" and \"flutter create .\" to regenerate."
```
Comment (xster) [inline packages/flutter_tools/bin/xcode_backend.sh:319]: ultra nit / optional: "Consider source controlling your project before regenerating to inspect the difference."

**Review by xster — COMMENTED** [review]
LGTM

**Comment by jmagman** [comment]
Boo, missed `macrobenchmarks`...

**Comment by jmagman** [comment]
> Boo, missed `macrobenchmarks`...

Never mind these tests are flaky.

**Comment by liyuqian** [comment]
[Some compile time regression][1] is detected after this commit. I'd like to confirm that this is expected due to build changes. I'd also like to know a little more about why it's increase significantly, and if there's any hope to make the compile a little faster. (The compile time is even slower when I introduced e2e in https://github.com/flutter/flutter/pull/65602). CC @jmagman @kf6gpe 

[1]: https://flutter-flutter-perf.skia.org/e/?begin=1598674983&end=1600234023&keys=Xff080f81588b83553db24d9cd652eb0a&requestType=0&xbaroffset=21337

**Comment by jmagman** [comment]
> [Some compile time regression](https://flutter-flutter-perf.skia.org/e/?begin=1598674983&end=1600234023&keys=Xff080f81588b83553db24d9cd652eb0a&requestType=0&xbaroffset=21337) is detected after this commit. I'd like to confirm that this is expected due to build changes. I'd also like to know a little more about why it's increase significantly, and if there's any hope to make the compile a little faster. (The compile time became even slower when I introduced e2e in #65602). CC @jmagman @kf6gpe

Before this PR the benchmarks were not thinning the Flutter frameworks to the targeted architecture.  https://github.com/flutter/flutter/pull/65124/files#diff-0d00f38df3883291f23cee8074ac075dL319-L320
They were calling:
```
BuildApp
EmbedFlutterFrameworks
```
instead of:
```
BuildApp
EmbedFlutterFrameworks
ThinAppFrameworks
```
The test benchmark app being transferred to the device should be much smaller now, which is a tradeoff we make in exchange for build time.  I'd have to look at the logs before and after to confirm that's the cause.  However as long as there isn't a compilation regression in a newly created/post v1.0 Flutter app, I'm not sure it's worth investigating too closely...

**Comment by liyuqian** [comment]
Sounds good. Thanks for the detailed explanation! We can move on as is.

**Pull request merged**